### ISSUE TITLE
Driver found the element but could not click because of the "An element command failed because the referenced element is no longer attached to the DOM." exception

### ISSUE THREAD

**Issue by tienletesting** [microsoft/WinAppDriver#1836]
When I tried to click on a hyperlink element, it throws the "An element command failed because the referenced element is no longer attached to the DOM." exception, even thought the element is displayed on the window that driver is focusing on.

The element is showing in Accessibility Inside as image below:
![image](https://user-images.githubusercontent.com/54089142/206820564-e1b3d5b5-2f20-4469-9206-56bda9e46f93.png)

Here is my code:
`var element = driver.FindElementByName("event#LinkId=1"); 
element.Click();` 

The exception was thrown at line 2.
The same exception was thrown when I tried to use MoveToElement() or element.Location.X.
But it works well if clicking on the parent element:
`driver.FindElementByAccessibilityId("linkTexts").Click();`

Another issue, the driver could not find the element if using:
`driver.FindElementByXPath("//Document[@AutomationId=\"linkTexts\"]/Hyperlink[contains(@Name, 'event')]"`

Below is some properties of the hyperlink element:
`ControlType = Hyperlink(50005)
IsEnabled	= True
IsOffscreen = False

IsKeyboardFocusable = False
HasKeyboardFocus	= False
`
Does anyone knows how to fix this issue? Thanks!

**Comment by Shakevg** [comment]
Another issue, the driver could not find the element if using:
driver.FindElementByXPath("//Document[@AutomationId=\"linkText\"]/Hyperlink[contains(@Name, 'event')]");
 => Save PageSource and check that this XPath is working via http://xpather.com/

Here is my code:
var element = driver.FindElementByName("event#LinkId=1");
element.Click();
=> What if you add some Sleep before it?

**Comment by tienletesting** [comment]
Thanks @Shakevg for your answer.

I tried on [http://xpather.com/]. 

No elements found if XPath is:
`//Document[@AutomationId="linkText"]/Hyperlink[contains(@**name**, 'event')]`

Or 
`//Document[@AutomationId="linkText"]/Hyperlink[contains(@**Name**, 'event')]`

Or 
`//Document[@AutomationId="linkText"]**//**Hyperlink[contains(@**Name**, 'event')]`

Or
`//Document[@AutomationId="linkText"]**//**Hyperlink[contains(@**name**, 'event')]`

**Elements found: 1** if XPath is:
`//Hyperlink[contains(@Name, 'event')]`  Or   `//Document[@AutomationId="linkText"]`

But when I applied `//Hyperlink[contains(@Name, 'event')]` into my code, it throws exception (even though the driver is desktop session)

![image](https://user-images.githubusercontent.com/54089142/207095564-ace49682-beb2-4184-a930-019bc7325c37.png)

Here is the form I was testing:
![image](https://user-images.githubusercontent.com/54089142/207096075-a0735ad1-64ea-4b8d-b0cb-f9ad24e2d84e.png)

After enter Name, I faced the problem.

_Here is my code:
var element = driver.FindElementByName("event#LinkId=1");
element.Click();
=> What if you add some Sleep before it?_
=> I added some Sleep before it but still get exception:
![image](https://user-images.githubusercontent.com/54089142/207097876-516f560c-2b28-492e-93c9-5fdf2f14665c.png)

Do you think the property `IsKeyboardFocusable = False` has something to do with those exceptions?

**Comment by tienletesting** [comment]
Does anyone has any idea about this issue?

I used try {} catch {} to debug the code as below:
`            var eventLink = DriverContext.Driver.FindElementByXPath("//*[contains(@Name,'event#')]");
            try
            {
                eventLink.Click();
            }
            catch (Exception ex)
            {
                if (ex is WebDriverException)
                {
                    eventLink = DriverContext.Driver.FindElementByXPath("//*[contains(@Name,'event#')]");
                    try
                    {
                        eventLink.Click();
                    }
                    catch (Exception ex1)
                    {
                        if (ex1 is WebDriverException)
                        {
                            eventLink = DriverContext.Driver.FindElementByXPath("//*[contains(@Name,'event#')]");
                        }
                        else
                            throw;
                    }
                }
                else
                    throw;
            }`

Every `DriverContext.Driver.FindElementByXPath("//*[contains(@Name,'event#')]")` step has the element found. No error.

But every `eventLink.Click();` step, it throws the "An element command failed because the referenced element is no longer attached to the DOM." exception

![image](https://user-images.githubusercontent.com/54089142/207129169-a4d8320e-c0d1-4b1a-ba72-9b494f294146.png)

**Comment by pm90pl** [comment]
@tienletesting could you provide WinAppDriver logs produced during that code execution?

**Comment by tienletesting** [comment]
@pm90pl Here is it:
`==========================================
POST /session/E107CB4F-8BAA-4E51-9C3F-CA90908A6D84/element HTTP/1.1
Accept: application/json, image/png
Connection: Keep-Alive
Content-Length: 48
Content-Type: application/json;charset=utf-8
Host: 192.168.101.125:4723
User-Agent: selenium/3.141.0 (.net windows)
X-Idempotency-Key: 49853f87-48ba-4d9d-a7e5-4fb24157c19f

{"using":"accessibility id","value":"linkTexts"}
HTTP/1.1 200 OK
Content-Length: 97
Content-Type: application/json

{"sessionId":"E107CB4F-8BAA-4E51-9C3F-CA90908A6D84","status":0,"value":{"ELEMENT":"42.27919650"}}


==========================================
POST /session/E107CB4F-8BAA-4E51-9C3F-CA90908A6D84/element/42.27919650/click HTTP/1.1
Accept: application/json, image/png
Connection: Keep-Alive
Content-Length: 2
Content-Type: application/json;charset=utf-8
Host: 192.168.101.125:4723
User-Agent: selenium/3.141.0 (.net windows)
X-Idempotency-Key: 981c1be2-bb42-43b8-824a-ae356f2e5681

{}
HTTP/1.1 200 OK
Content-Length: 63
Content-Type: application/json

{"sessionId":"E107CB4F-8BAA-4E51-9C3F-CA90908A6D84","status":0}


==========================================
POST /session/E107CB4F-8BAA-4E51-9C3F-CA90908A6D84/element HTTP/1.1
Accept: application/json, image/png
Connection: Keep-Alive
Content-Length: 57
Content-Type: application/json;charset=utf-8
Host: 192.168.101.125:4723
User-Agent: selenium/3.141.0 (.net windows)
X-Idempotency-Key: b2f66ae7-707e-46a0-a91e-9dd883cf3de8

{"using":"xpath","value":"//*[contains(@Name,'event#')]"}
HTTP/1.1 200 OK
Content-Length: 101
Content-Type: application/json

{"sessionId":"E107CB4F-8BAA-4E51-9C3F-CA90908A6D84","status":0,"value":{"ELEMENT":"42.27919650.4.0"}}


==========================================
POST /session/E107CB4F-8BAA-4E51-9C3F-CA90908A6D84/element/42.27919650.4.0/click HTTP/1.1
Accept: application/json, image/png
Connection: Keep-Alive
Content-Length: 2
Content-Type: application/json;charset=utf-8
Host: 192.168.101.125:4723
User-Agent: selenium/3.141.0 (.net windows)
X-Idempotency-Key: d50dfa71-96a1-447b-8ac7-e200690feb25

{}
HTTP/1.1 400 Bad Request
Content-Length: 160
Content-Type: application/json

{"status":10,"value":{"error":"stale element reference","message":"An element command failed because the referenced element is no longer attached to the DOM."}}


==========================================
POST /session/E107CB4F-8BAA-4E51-9C3F-CA90908A6D84/element HTTP/1.1
Accept: application/json, image/png
Connection: Keep-Alive
Content-Length: 57
Content-Type: application/json;charset=utf-8
Host: 192.168.101.125:4723
User-Agent: selenium/3.141.0 (.net windows)
X-Idempotency-Key: 10fd3420-72ac-48f7-b8ea-bd56620e2598

{"using":"xpath","value":"//*[contains(@Name,'event#')]"}
HTTP/1.1 200 OK
Content-Length: 101
Content-Type: application/json

{"sessionId":"E107CB4F-8BAA-4E51-9C3F-CA90908A6D84","status":0,"value":{"ELEMENT":"42.27919650.4.0"}}


==========================================
POST /session/E107CB4F-8BAA-4E51-9C3F-CA90908A6D84/element/42.27919650.4.0/click HTTP/1.1
Accept: application/json, image/png
Connection: Keep-Alive
Content-Length: 2
Content-Type: application/json;charset=utf-8
Host: 192.168.101.125:4723
User-Agent: selenium/3.141.0 (.net windows)
X-Idempotency-Key: b1303d8b-6f86-4cd8-844f-32a466c9d264

{}
HTTP/1.1 400 Bad Request
Content-Length: 160
Content-Type: application/json

{"status":10,"value":{"error":"stale element reference","message":"An element command failed because the referenced element is no longer attached to the DOM."}}
`

**Comment by pm90pl** [comment]
@tienletesting could you also provide 'PageSource'?

**Comment by TienMS** [comment]
@pm90pl here is it:

![image](https://user-images.githubusercontent.com/120534997/207655863-69212381-b728-4b13-b6d5-f4dbd0d20e49.png)

![image](https://user-images.githubusercontent.com/120534997/207655235-3d8625e5-1a78-4eb6-9340-3283bd308c6f.png)

**Comment by tienletesting** [comment]
@pm90pl here is it:
 
![image](https://user-images.githubusercontent.com/54089142/207656224-5ec19db0-542e-40ab-9856-ac2791f3a1c3.png)

![image](https://user-images.githubusercontent.com/54089142/207656269-8b3c6808-f753-4b8d-9b78-750dbfe704d6.png)

**Comment by pm90pl** [comment]
@tienletesting ok, can't say anything right now. Is it the only element that can be found using given XPath? Can you perform any other operation on found element? Could you try to take w screenshot of the element using WinAppDriver?

**Comment by tienletesting** [comment]
@pm90pl I am sure it is the only element that can be found using the given XPath. I could not perform any other operation on the found element. All operation likes Click, get Location/Size, MoveToElement... throws ""An element command failed because the referenced element is no longer attached to the DOM." exception".

I could not take screenshot for that specific element because the exception throws when trying to get location of the element.

`            var linkTexts = DriverContext.Driver.FindElementByAccessibilityId("linkTexts");`
`            var eventLink = linkTexts.FindElementByXPath("//*[contains(@Name,'event#LinkId')]");`
 `           TakeScreenshot(DriverContext.Driver, eventLink);`

![image](https://user-images.githubusercontent.com/54089142/207935463-2b890789-bc6b-496a-ba10-5fb15c2855cf.png)

**Comment by anunay1** [comment]
What is shown in inepect.exe?

**Comment by anunay1** [comment]
Please use inspect.exe and share what is  shown as it has more information.

**Comment by tienletesting** [comment]
@anunay1 Here
![image](https://user-images.githubusercontent.com/54089142/208143654-5b184107-c753-4014-acc0-cd690a81859d.png)

**Comment by anunay1** [comment]
The name is dynamic, with the below code try to get the child elements first: 
driver.findelementbyname("document").findelementsbyxpth("*/*")

and check if you get any element.

**Comment by tienletesting** [comment]
No, I dont get any element. "linkTexts" is AutomationID of the "document" tag.

![image](https://user-images.githubusercontent.com/54089142/208156710-c3e31744-fb63-43b6-96ea-a489c5aa4281.png)

If I use `FindElementByXPath("//*[contains(@Name,'event#LinkId')]")`, then I get one element:

![image](https://user-images.githubusercontent.com/54089142/208157276-46db49c3-7bc2-436e-8c63-7da273ab58e3.png)

If I use `FindElementByXPath("/*[contains(@Name,'event#LinkId')]")`, then exception thrown:

![image](https://user-images.githubusercontent.com/54089142/208157561-5cb7d7f6-beaf-4a8a-8065-e999399be106.png)

**Comment by anunay1** [comment]
Can you try driver.findelementbyautomatioid("linkTexts").findelementsbytagname("Link")

**Comment by tienletesting** [comment]
@anunay1 No. No elements found if using `driver.findelementbyautomatioid("linkTexts").FindElementsByTagName("Link")` or `driver.findelementbyautomatioid("linkTexts").FindElementsByTagName("Hyperlink")`

![image](https://user-images.githubusercontent.com/54089142/208160527-6a64e605-4d45-443a-b323-32e7bd764bbe.png)

**Comment by anunay1** [comment]
That's really weird. Can you ping me on teams?

**Comment by tienletesting** [comment]
@anunay1 Yes, that's weird. I can't. I am sorry. Do you need any other information?

**Comment by liljohnak** [comment]
@tienletesting like this?
`driver.findelementbyname("document").findelementsbyxpth("*/*")`

**Comment by tienletesting** [comment]
@liljohnak Yes, it found 2 elements as expected.

![image](https://user-images.githubusercontent.com/54089142/208182785-621dfcab-a3a8-4fc8-a6be-95955329b4be.png)

**Comment by perfectprocrastinator** [comment]
I am facing the same issue ,using WinAppDriver to automate Excel.
Is there any update on this

**Comment by anunay1** [comment]
> I am facing the same issue ,using WinAppDriver to automate Excel. Is there any update on this

You need to retry. That's the only way out.
